### PR TITLE
Closures are recursively reachable

### LINKED ISSUE #126012: Missing optimized MIR for an item
```console
$ cat a.rs 
pub static A: fn() -> fn() = {
    #[inline(never)]
    fn f() {}
    #[inline(always)]
    || -> fn() { f }
};
$ cat b.rs
static A: fn() -> fn () = a::A;
fn main() { A()() }
$ rustc a.rs --crate-type=lib
$ rustc b.rs --extern a -L.
error: missing optimized MIR for an item in the crate `a`
  |
note: missing optimized MIR for this item (was the crate `a` compiled with `--emit=metadata`?)
 --> /.../a.rs:3:5
  |
3 |     fn f() {}
  |     ^^^^^^

error: aborting due to 1 previous error
```

Regression in c563f2ee799b285067124b516ce99f26063f8351 (#122371 cc @oli-obk @tmiasko).

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

### PR DESCRIPTION
**Pull request by tmiasko**
Fixes #126012.

### CODE REVIEW

File: compiler/rustc_passes/src/reachable.rs [hunk 1]
```diff
@@ -157,6 +157,7 @@ impl<'tcx> ReachableContext<'tcx> {
                 }
                 hir::ImplItemKind::Type(_) => false,
             },
+            Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(..), .. }) => true,
```
Comment (compiler-errors) [inline compiler/rustc_passes/src/reachable.rs:160]: Why isn't this:

```suggestion
            Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(hir::Closure { def_id, .. }), .. }) => recursively_reachable(def_id),
```

?

**Comment by compiler-errors** [comment]
Also is there a bug report or anything about this issue? Seems strange to have such little information attached to this PR 😅

**Comment by apiraino** [comment]
Sorry this backport was accepted last week ([Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-06-06/near/443071818)) but I missed to mark it as such. 

@rustbot label +beta-accepted

**Comment by cuviper** [comment]
It's too late for 1.79.0, but this could be stable-nominated if needed.

@rustbot label -beta-nominated -beta-accepted

**Comment by apiraino** [comment]
ok, then I'll stable nominate since there's another PR candidate for a stable backport (if a dot release will be considered)

@rustbot +stable-nominate

**Comment by apiraino** [comment]
Stable backport accpeted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-06-20/near/445864816), if a point release is planned.

 A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +stable-accepted

**Comment by cuviper** [comment]
Clearing stable backport labels since we didn't ship a 1.79 point release, and this is already in current stable 1.80.0.

@rustbot label -stable-nominated -stable-accepted

**Pull request merged**